consensus/istanbul/validator.go

```diff
@@ -36,6 +37,31 @@ type ValidatorData struct {
 	BLSPublicKey blscrypto.SerializedPublicKey
 }
 
+type ValidatorDataWithCache struct {
+	Address      common.Address
+	BLSPublicKey blscrypto.SerializedPublicKey
+	Uncompressed []byte
+}
+
+func (v *ValidatorDataWithCache) UnmarshalJSON(data []byte) (err error) {
+	tmp := struct {
+		Address      common.Address
+		BLSPublicKey blscrypto.SerializedPublicKey
+		Uncompressed []byte
+	}{}
+	err = json.Unmarshal(data, &tmp)
+	if err != nil {
+		return err
+	} else if tmp.Uncompressed == nil && tmp.BLSPublicKey != (blscrypto.SerializedPublicKey{}) {
+		return fmt.Errorf("Cached key missing")
```

Comment:
Will this "get fixed" later? Will anyone catch this error and try to re-initialize it correctly?
If not, maybe it's better to directly fill in the uncompressed part?

Reply:
It will be caught when snapshot is loaded. It's better to error here, that way the old snapshot will get replaced with the one that has the uncompressed key.

Reply:
Can you show where the retry happens? I had a hard time tracing it

Reply:
It tries to load if here https://github.com/celo-org/celo-blockchain/blob/mrsmkl/precompile-bls-key/consensus/istanbul/backend/engine.go#L861
If it fails, it will create a new one eventually.

Reply:
Moved the check to a different place.